### ISSUE TITLE
Change mark type of initial scatter plot (to get bar chart or other type of plot)

### ISSUE THREAD

**Issue by eharkins** [matsengrp/olmsted#41]
We want to allow the user to change the mark type in a drop-down menu. Changing the vega mark type means essentially getting a different style of plot such as a bar chart, box plot, etc. This would probably be more than just allowing for a change of mark-type, as the data, axes, etc have to be compatible with the chosen mark-type for it to create a viz that makes sense / works.

**Comment by eharkins** [comment]
@metasoarous we should talk about what sort of mark types we want to allow, based on what would actually be useful to see given the data. After determining this, it may be the case that well we want to restrict the pairings of mark types with certain summary statistics that don't make sense together if those pairings exist based on what mark types we think might be helpful and which summary statistics we are including in the axis dropdowns

**Comment by metasoarous** [comment]
I think we might want
* [ ] bar chart (with binning as appropriate)
* [ ] box and wiskers (binning as appropriate)
* [ ] violin (binning as appropriate)
* [ ] pie charts showing timepoint distribution
* [ ] density contours (https://vega.github.io/vega/examples/contour-plot/)
* [ ] lines connecting clusters that are actually from the same clonal family?

You're right that in many of these cases we'll have to be restricting the variables assignable for one or the other of the axes, and in other cases responding differently depending on which kind of variable is selected (e.g. binning necessary for bars etc on numeric x-axis, but not for nominal values). This seems to suggest that we'll have some top level configuration selector, and may compose the vega a little differently in each case.

**Comment by eharkins** [comment]
Sounds good. I wonder how expensive binning is and if we can create the various binned data sources we need in the vega data specification, which would mean just changing the data source when selecting a new mark type as opposed to doing binning in that moment.

**Comment by metasoarous** [comment]
Vega has pretty efficient binning. It should just be a change in spec.

**Comment by eharkins** [comment]
So we can do easy manipulation (binning and creating separate data sources) of the data, and we already know we can set the domain and range of the scales of our viz to be variables (aka signals). One thing that doesn't seem obvious is how to change the `type` field of the scales. The case is that if we switch the mark type for our viz, we might be switching to a different type of plot that requires a different `type` of scale. The ways I can think to do this are:
- change the `type` somehow dynamically through a signal
- change the scale used by the axes and have a variety of scales to choose from in our spec
- change the spec itself (through loading a separate component?)if there is no way to change the type of a scale or to switch scales within a spec
- stick to only using one type of scale - maybe changing scale types is a feature we don't need, but based on the above suggested mark types, it seems like it will be.

**Comment by eharkins** [comment]
The more I think about it the more it seems there is no reason not to do the third option mentioned above for our own sanity, in terms of keeping specs to a reasonable size and just switching them out when that is appropriate. I'm open to an argument for why it should all be the same spec that is ultra-configurable, but at this level it seems as thought we could have different versions of the same spec load through a React component based on props passed to it, without the UX suffering.

**Comment by metasoarous** [comment]
This is a very pragmatic approach, and probably the fastest route to victory.

The main reason for not doing it this way is portability. It would be great to be able to copy the entire vega spec and put it in something that takes vega (vega editor for example; embed in a scientists blog; etc) and have it completely interactive. I think you have the right idea about doing it all in vega, and maybe it wouldn't be too bad, but its still likely to be more work figuring out how, whereas we have a clear path forward for doing it in react. Only other reason I can think of to do it all in vega is that you'd probably learn a ton doing it and could do more fine grained things with vega in the future.

**Comment by eharkins** [comment]
Good point about being able to share the spec / have it all in vega. I've tried briefly to do the first two options I suggested (which are the only two I could think of to keep this within the spec) without immediate success:
> * change the `type` somehow dynamically through a signal
> * change the scale used by the axes and have a variety of scales to choose from in our spec

It feels like there should be a way to change the mark type if the data will allow for it, and may even be an example out there somewhere. Maybe a place to start is in the voyager app where you can change the mark type from the sidebar menu; how are they changing the marktype of the spec? Or are they just creating a new spec? Let me know if you have other suspicions about how this could be possible, @metasoarous, because I agree it would be nice to have that be a part of the spec.

**Comment by metasoarous** [comment]
They are probably just doing it outside the spec, since the dropdowns aren't immediately next to the viz.

I think you have your head in the right place with this. If you could create signals for the type and other config information, you'd be able to dynamically get everything piped in.

**Comment by metasoarous** [comment]
I've been thinking about this, and think this might be better as a v1-bonus issue. I think there are some other issues which are at least on the same order or magnitude of usefulness, but which might be much easier to implement.
